File: 2024/day-02/main.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>

#define MAX_REPORT_LENGTH 1000

// Function to check if a report is safe
bool is_safe(int *report, int size) {
    bool all_increasing = true, all_decreasing = true;

    for (int i = 0; i < size - 1; i++) {
        int diff = report[i + 1] - report[i];
        if (abs(diff) < 1 || abs(diff) > 3) {
            return false;
        }
        if (diff <= 0) all_increasing = false;
        if (diff >= 0) all_decreasing = false;
    }

    return all_increasing || all_decreasing;
}

// Function to check if a report can be made safe by removing one level
bool can_be_made_safe(int *report, int size) {
    for (int i = 0; i < size; i++) {
        int modified[MAX_REPORT_LENGTH];
        int k = 0;

        // Create a modified report without the i-th level
        for (int j = 0; j < size; j++) {
            if (j != i) {
                modified[k++] = report[j];
            }
        }

        if (is_safe(modified, size - 1)) {
            return true;
        }
    }
    return false;
}

// Function to count safe reports
int count_safe_reports(const char *file_path, int part) {
    FILE *file = fopen(file_path, "r");
    if (!file) {
        fprintf(stderr, "Error: Could not open file %s\n", file_path);
        return -1;
    }

    char line[1024];
    int safe_count = 0;

    while (fgets(line, sizeof(line), file)) {
        if (line[0] == '\n' || line[0] == '\0') continue;

        int report[MAX_REPORT_LENGTH];
        int size = 0;

        char *token = strtok(line, " ");
        while (token) {
            report[size++] = atoi(token);
            token = strtok(NULL, " ");
        }

        if (part == 1) {
            if (is_safe(report, size)) safe_count++;
        } else if (part == 2) {
            if (is_safe(report, size) || can_be_made_safe(report, size)) {
                safe_count++;
            }
        }
    }

    fclose(file);
    return safe_count;
}

int main() {
    const char *file_path = "input.txt";

    // Part 1
    int part1 = count_safe_reports(file_path, 1);
    printf("Part 1: Number of safe reports: %d\n", part1);

    // Part 2
    int part2 = count_safe_reports(file_path, 2);
    printf("Part 2: Number of safe reports: %d\n", part2);

    return 0;
}

File: 2024/day-02/main.ex
defmodule ReactorSafety do
  def is_safe(report) do
    differences = Enum.chunk_every(report, 2, 1, :discard) |> Enum.map(fn [a, b] -> b - a end)

    valid_differences = Enum.all?(differences, fn diff -> diff in -3..-1 or diff in 1..3 end)
    all_increasing = Enum.all?(differences, &(&1 > 0))
    all_decreasing = Enum.all?(differences, &(&1 < 0))

    valid_differences and (all_increasing or all_decreasing)
  end

  def can_be_made_safe(report) do
    Enum.any?(0..(length(report) - 1), fn i ->
      modified = List.delete_at(report, i)
      is_safe(modified)
    end)
  end

  def count_safe_reports(file_path, part) do
    file_path
    |> File.stream!()
    |> Enum.map(&String.trim/1)
    |> Enum.reject(&(&1 == ""))
    |> Enum.map(fn line -> String.split(line) |> Enum.map(&String.to_integer/1) end)
    |> Enum.count(fn report ->
      case part do
        1 -> is_safe(report)
        2 -> is_safe(report) or can_be_made_safe(report)
      end
    end)
  end
end

# Main program
file_path = "input.txt"

part1 = ReactorSafety.count_safe_reports(file_path, 1)
IO.puts("Part 1: Number of safe reports: #{part1}")

part2 = ReactorSafety.count_safe_reports(file_path, 2)
IO.puts("Part 2: Number of safe reports: #{part2}")
